### ISSUE TITLE
Quantization FP16 model using pytorch_quantization and TensorRT 

### ISSUE THREAD

**Issue by WoodieDudy** [NVIDIA/TensorRT#3219]
**Objective:**
My primary goal is to accelerate my model's performance using int8 + fp16 quantization. To achieve this, I first need to quantize the model and then calibrate it.
As far as I understand, there are two quantization methods available:

1. **Implicit Quantization**: Using `trtexec`
2. **Explicit Quantization**: Using `pytorch_quantization`

**Questions/Concerns:**

1. How can I achieve int8 + fp16 model representation using explicit quantization with `pytorch_quantization`? From what I gathered, the current release does not support fp16 quantization of a model as mentioned [here](https://github.com/NVIDIA/TensorRT/issues/3095#issuecomment-1672980900).

2. Is calibration directly supported in fp16 through `pytorch_quantization`?
   
3. Is calibration in fp16 supported through `trtexec`?
   
4. If it's not yet supported, do you have any suggestions or workarounds to force kernels, input/output formats to be fp16 after calibration using fp32? Moreover, how can I set the data type for the inputs and outputs of all layers, especially if I have a large model where manual configuration is not feasible? Is this achievable through flags for `trtexec`, or is building the engine from Python more recommended?

**Code Examples**:
SVGs are created by this script
```bash
python TensorRT/tools/experimental/trt-engine-explorer/utils/process_engine.py model.onnx folder int8 fp16
```

*Explicit Quantization using pytorch_quantization*:
```python
from pytorch_quantization import nn as quant_nn
from pytorch_quantization import quant_modules

quant_nn.TensorQuantizer.use_fb_fake_quant = True
quant_modules.initialize()            

model = nn.Linear(512, 2048)

torch.onnx.export(
    model.to(dtype=torch.float32, device='cuda'),
    torch.rand(1024, 512).to(dtype=torch.float32, device='cuda'),
    save_dir,
    verbose=False,
    input_names=["x"]
)

```
![QKV_explicit_quant_fp32 onnx engine graph json](https://github.com/NVIDIA/TensorRT/assets/55787608/5cb21c66-c02d-4976-b04c-265b70b1e9e9)

*Implicit Quantization*:
```python
model = nn.Linear(512, 2048)

torch.onnx.export(
    model.to(dtype=torch.float16, device='cuda'),
    torch.rand(1024, 512).to(dtype=torch.float16, device='cuda'),
    save_dir,
    verbose=False,
    input_names=["x"]
)

```
![QKV_int8_real_fp16](https://github.com/NVIDIA/TensorRT/assets/55787608/d4b827c0-44f1-4a72-b5a2-8e4ae72548c8)

Any assistance or suggestions would be greatly appreciated.

**Comment by zerollzeng** [comment]
> * How can I achieve int8 + fp16 model representation using explicit quantization with `pytorch_quantization`? From what I gathered, the current release does not support fp16 quantization of a model as mentioned [here](https://github.com/NVIDIA/TensorRT/issues/3095#issuecomment-1672980900).
> *

For FP16 you don't need calibration, you can add Q/DQ on the wanted layers to let it run in INT8, and other layers will running in FP16/FP32.

> 4\. If it's not yet supported, do you have any suggestions or workarounds to force kernels, input/output formats to be fp16 after calibration using fp32? Moreover, how can I set the data type for the inputs and outputs of all layers, especially if I have a large model where manual configuration is not feasible? Is this achievable through flags for `trtexec`, or is building the engine from Python more recommended?

Check trtexec -h, you will find the error to set layer precision and intermediate tensor precision.

**Comment by vadimkantorov** [comment]
> For FP16 you don't need calibration, you can add Q/DQ on the wanted layers to let it run in INT8, and other layers will running in FP16/FP32.

@WoodieDudy meant calibrating a quantized int8 model which is based on the original fp16 model and uses fp16 for inputs and for some intermediate ops. Currently, `pytorch_quantization` seems to not support float16 as dtype, as @ttyio suggested at https://github.com/NVIDIA/TensorRT/issues/3095#issuecomment-1673516602 .

The question is what is the typical flow now for the usecase:
1) original model in fp16
2) we would like to make certain nodes to run in int8 (with fp16 as intermediate dtype)
3) we would like to calibrate the model (using pytorch_quantization? or by trtexec? tensorrt directly?)
4) we would like to run the resulting int8+fp16 model in TensorRT

How can we do the calibration of the qparams for this very typical usecase? Thank you :)

Do I understand correctly that for `pytorch_quantization` we should currently do calibration in FP32 copy of the model and then hope for the best that using these qparams with fp16 inference isn't going to affect accuracy? Or is there a better way to work around absence of support of fp16 in pytorch_quantization?

**Comment by vadimkantorov** [comment]
@ttyio The question is about calibration of the int8 model parts (and not about running in fp16 per se), given that we want to keep most of intermediate tensors in fp16. In this setup, the calibration / FakeQuant should also be happening in fp16, right?

As I understood from our discussion, this is currently not supported natively by `pytorch_quantization`. So I am wondering what is the recommended solution for the moment? Do calibration using the fp32 version of the model? And then just have TensorRT use the estimated qparams while also changing from fp32->fp16?

**Comment by ttyio** [comment]
@vadimkantorov , In the original question I see both implicit quantization and explicit quantization choose INT8->FP16 kernel, and we insert FP16->Fp32 reformat only in the explicit quantization, and if this is the concern, we can change the network IO data type to remove this reformat. It's not related to calibration itself. For a sequence of INT8 operators, the efficient way is to place all the intermediate tensors in int8. Even for the mix precision training (with fp16 enabled) https://pytorch.org/docs/stable/amp.html, the export onnx is still fp32, and trt could internally optimize the perf by selecting fp16.

**Comment by vadimkantorov** [comment]
Well, the implicit quantization works with fp16 but we have control only via `trtexec` flags (we've only used the most basic `--int8` flag, did not attempt any calibration with `trtexec`).

The explicit quantization lets us do do faithful calibration, but only in fp32 during calibration (without accounting for smaller precision at inference). So we are wondering what are best practices for doing qparams (hopefully faithful) calibration while also using fp16 ops in inference.

We do understand that we can force `trtexec`/tensorrt to select fp16 op in practice (even when the export is using only fp32). We are wondering about calibration specifically.

**Comment by ttyio** [comment]
@vadimkantorov , could you elaborate more on the motivation for calibration using fp16 + int8 and export fp16 + int8 onnx instead of fp32 + int8 onnx?

Previously we treat amp and qat as orthographic features, we have a test where they work together and export as fp32 onnx: https://github.com/NVIDIA/TensorRT/blob/release/8.6/tools/pytorch-quantization/tests/integration_test.py#L81

**Comment by ttyio** [comment]
closing since no activity for more than 3 weeks, pls reopen if you still have question. thanks!
